mtsp/src/main.leo

```diff
@@ -0,0 +1,673 @@
+  transition transfer_from_public(token_id: field, owner: address, recipient: address, amount: u128) {
+    return then finalize(token_id, owner, recipient, amount, self.caller);
+  }
+
+  finalize transfer_from_public(token_id: field, owner: address, recipient: address, amount: u128, spender: address) {
+    // Check that the spender is authorized to spend the amount
+    let allowance: Allowance = Allowance {
+      account: owner,
+      spender: spender,
+      token_id: token_id
+    };
+    let allowance_key: field = BHP256::hash_to_field(allowance);
+    let current_allowance: u128 = allowances.get(allowance_key);
+    assert(current_allowance >= amount);
+    // Update the allowance
+    allowances.set(allowance_key, current_allowance - amount);
+
+    // Check that the owner has enough authorized balance
+    let owner_key: TokenOwner = TokenOwner {
+      account: owner,
+      token_id: token_id
+    };
+    let owner_key_hash: field = BHP256::hash_to_field(owner);
+    let balance: Balance = authorized_balances.get(owner_key_hash);
+    assert_eq(owner, balance.account);
+    assert(balance.balance >= amount);
+
+    // Update the balance for the owner
+    let new_balance: Balance = Balance {
+      token_id: token_id,
+      account: account,
+      balance: balance.balance - amount,
+      authorized_until: balance.authorized_until
+    };
+    authorized_balances.set(owner_key_hash, new_balance);
+
+    // Get or create the balance for the recipient
+    let recipient_key: TokenOwner = TokenOwner {
+      account: recipient,
+      token_id: token_id
+    };
+    let recipient_balance_key: field = BHP256::hash_to_field(recipient_key);
+    // Get the locked balance if authorization is required, otherwise get the authorized balance
+    let authorization_required: bool = registered_tokens.get(token_id).external_authorization_required;
+    let default_expiration: u32 = authorization_required ? 0u32 : 4294967295u32;
+    let default_balance: Balance = Balance {
+      token_id: token_id,
+      account: recipient,
+      balance: 0u128,
+      authorized_until: default_expiration
+    };
+    let recipient_balance: Balance = authorization_required ? balances.get_or_use(recipient_balance_key, default_balance) : authorized_balances.get_or_use(recipient_balance_key, default_balance);
+    let new_recipient_balance: Balance = Balance {
+      token_id: token_id,
+      account: recipient,
+      balance: recipient_balance.balance + amount,
+      authorized_until: balance.authorized_until
+    };
+    // Update the balance of the recipient
+    if (authorization_required) {
+      balances.set(recipient_balance_key, new_recipient_balance);
+    } else {
+      authorized_balances.set(recipient_balance_key, new_recipient_balance);
+    }
+  }
+
+  transition transfer_public_to_private(token_id: field, recipient: address, amount: u128, preauthorize_transfer_record: bool) -> Token {
```

Comment:
To call this method from another program, it must know which type of token it is dealing with. The only way to do that within the transition is to pass this argument from the initial invocation by the EOA. This can make things very complicated on the frontend side if we are dealing with a non-trivial chain of calls, especially if there are multiple calls to this method in the chain.